extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -547,118 +579,167 @@ private void updatePartitionDataFromKafka()
-                  log.warn("Stopping task [%s] which does not match the expected partition allocation", taskId);
-                  stopTask(taskId, false);
-                  taskGroupId = null;
-                  break;
-                }
-              }
-
-              if (taskGroupId == null) {
-                continue;
-              }
-
-              if (!taskGroups.containsKey(taskGroupId)) {
-                log.debug("Creating new task group [%d]", taskGroupId);
-                taskGroups.put(
-                    taskGroupId,
-                    new TaskGroup(
-                        kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap(),
-                        kafkaTask.getIOConfig().getMinimumMessageTime()
-                    )
-                );
-              }
+      if (!(task instanceof KafkaIndexTask) || !dataSource.equals(task.getDataSource())) {
+        continue;
+      }
 
-              if (!isTaskCurrent(taskGroupId, taskId)) {
-                log.info("Stopping task [%s] which does not match the expected parameters and ingestion spec", taskId);
-                stopTask(taskId, false);
-              } else {
-                taskGroups.get(taskGroupId).tasks.put(taskId, new TaskData());
-              }
-            }
-          }
+      taskCount++;
+      final KafkaIndexTask kafkaTask = (KafkaIndexTask) task;
+      final String taskId = task.getId();
+
+      // Determine which task group this task belongs to based on one of the partitions handled by this task. If we
+      // later determine that this task is actively reading, we will make sure that it matches our current partition
+      // allocation (getTaskGroupIdForPartition(partition) should return the same value for every partition being read
+      // by this task) and kill it if it is not compatible. If the task is instead found to be in the publishing
+      // state, we will permit it to complete even if it doesn't match our current partition allocation to support
+      // seamless schema migration.
+
+      Iterator<Integer> it = kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet().iterator();
+      final Integer taskGroupId = (it.hasNext() ? getTaskGroupIdForPartition(it.next()) : null);
+
+      if (taskGroupId != null) {
+        // check to see if we already know about this task, either in [taskGroups] or in [pendingCompletionTaskGroups]
+        // and if not add it to taskGroups or pendingCompletionTaskGroups (if status = PUBLISHING)
+        TaskGroup taskGroup = taskGroups.get(taskGroupId);
+        if (!isTaskInPendingCompletionGroups(taskId) && (taskGroup == null || !taskGroup.tasks.containsKey(taskId))) {
+
+          futureTaskIds.add(taskId);
+          futures.add(
+              Futures.transform(
+                  taskClient.getStatusAsync(taskId), new Function<KafkaIndexTask.Status, Boolean>()
+                  {
+                    @Nullable
```

Comment:
Guessing this was boilerplate added by IntelliJ; it looks like this won't return null.